drivers/SmartThings/zigbee-smoke-detector/src/test/test_aqara_gas_detector.lua

```diff
@@ -0,0 +1,212 @@
+    cluster_base.write_manufacturer_specific_attribute(mock_device, PRIVATE_CLUSTER_ID,
+    PRIVATE_MUTE_ATTRIBUTE_ID, MFG_CODE, data_types.Uint8, 1) })
+  end
+)
+
+
+
+test.register_coroutine_test(
+  "selfCheck report should be handled",
+  function()
+    local attr_report_data = {
+      { PRIVATE_SELF_CHECK_ATTRIBUTE_ID, data_types.Uint8.ID, 0x01 }
+    }
+    test.socket.zigbee:__queue_receive({
+      mock_device.id,
+      zigbee_test_utils.build_attribute_report(mock_device, PRIVATE_CLUSTER_ID, attr_report_data, MFG_CODE)
+    })
+    test.socket.capability:__expect_send(mock_device:generate_test_message("main",
+    selfCheck.selfCheckState.selfCheckCompleted()))
+  end
+)
+
+
+
+test.register_coroutine_test(
+  "Capability on command should be handled : device selfCheck",
+  function()
+    test.socket.capability:__queue_receive({ mock_device.id,
+      { capability = selfCheckId, component = "main", command = "startSelfCheck", args = {state = "selfCheckCompleted"} } })
+    test.socket.capability:__expect_send(mock_device:generate_test_message("main",selfCheck.selfCheckState.selfChecking()))
+    test.socket.zigbee:__expect_send({ mock_device.id,
+    cluster_base.write_manufacturer_specific_attribute(mock_device, PRIVATE_CLUSTER_ID,
+    PRIVATE_SELF_CHECK_ATTRIBUTE_ID, MFG_CODE, data_types.Boolean, true) })
+  end
+)
+
+
+
+test.register_coroutine_test(
+  "lifetime report should be handled",
+  function()
+    local attr_report_data = {
+      { PRIVATE_LIFE_TIME_ATTRIBUTE_ID, data_types.Uint8.ID, 0x01 }
+    }
+    test.socket.zigbee:__queue_receive({
+      mock_device.id,
+      zigbee_test_utils.build_attribute_report(mock_device, PRIVATE_CLUSTER_ID, attr_report_data, MFG_CODE)
+    })
+    test.socket.capability:__expect_send(mock_device:generate_test_message("main",
+    lifeTimeReport.lifeTimeState.endOfLife()))
+  end
+)
+
+
+
+test.register_coroutine_test(
+  "sensitivityAdjustment report should be handled",
+  function()
+    local attr_report_data = {
+      { PRIVATE_SENSITIVITY_ADJUSTMENT_ATTRIBUTE_ID, data_types.Uint8.ID, 0x01 }
+    }
+    test.socket.zigbee:__queue_receive({
+      mock_device.id,
+      zigbee_test_utils.build_attribute_report(mock_device, PRIVATE_CLUSTER_ID, attr_report_data, MFG_CODE)
+    })
+    test.socket.capability:__expect_send(mock_device:generate_test_message("main",
+    sensitivityAdjustment.sensitivityAdjustment.Low()))
+  end
+)
+
+
+
+test.register_coroutine_test(
+  "Capability on command should be handled : setSensitivityAdjustment Low",
+  function()
+    test.socket.capability:__queue_receive({ mock_device.id,
+      { capability = sensitivityAdjustmentId, component = "main", command = "setSensitivityAdjustment", args = {"Low"}}} )
+    test.socket.zigbee:__expect_send({ mock_device.id,
+    cluster_base.write_manufacturer_specific_attribute(mock_device, PRIVATE_CLUSTER_ID,
+    PRIVATE_SENSITIVITY_ADJUSTMENT_ATTRIBUTE_ID, MFG_CODE, data_types.Uint8, 0x01) })
```

Comment:
@Hongming6 

> Can you write a test for this behavior?
> 
> You can expand your existing test that sets the sensitivity by adding a report from the device followed by another set to the same value. You would expect to see an immediate event.

I think green means you add a report from device here, then send setSensitivityAdjustment command again
```
    test.socket.capability:__queue_receive({ mock_device.id,
      { capability = sensitivityAdjustmentId, component = "main", command = "setSensitivityAdjustment", args = {"Low"}}} )
```
But this time driver should send `sensitivityAdjustment.sensitivityAdjustment.Low()` directly instead of sendding zigbee message.
```
    test.socket.capability:__expect_send(mock_device:generate_test_message("main",
    sensitivityAdjustment.sensitivityAdjustment.Low()))
```

Reply:
@greens I roughly understand what you mean. I tried to modify the code, but there was an error. Am I doing this right?
`test.register_coroutine_test(
  "Capability on command should be handled : setSensitivityAdjustment Low",
  function()
    local attr_report_data = {
      { PRIVATE_SENSITIVITY_ADJUSTMENT_ATTRIBUTE_ID, data_types.Uint8.ID, 0x01 }
    }
    test.socket.capability:__queue_receive({ mock_device.id,
      { capability = sensitivityAdjustmentId, component = "main", command = "setSensitivityAdjustment", args = {"Low"}}} )
    test.socket.zigbee:__expect_send({ mock_device.id,
      cluster_base.write_manufacturer_specific_attribute(mock_device, PRIVATE_CLUSTER_ID,
      PRIVATE_SENSITIVITY_ADJUSTMENT_ATTRIBUTE_ID, MFG_CODE, data_types.Uint8, 0x01) })
    test.socket.zigbee:__queue_receive({
      mock_device.id,
      zigbee_test_utils.build_attribute_report(mock_device, PRIVATE_CLUSTER_ID, attr_report_data, MFG_CODE)
    })
    test.socket.capability:__queue_receive({ mock_device.id,
      { capability = sensitivityAdjustmentId, component = "main", command = "setSensitivityAdjustment", args = {"Low"}}} )
    test.socket.capability:__expect_send(mock_device:generate_test_message("main",
      sensitivityAdjustment.sensitivityAdjustment.Low()))
  end
)`

The error information is as follows

`DEBUG || zigbee-smoke-detector device thread event handled
TRACE || Received event with handler capability
INFO  || <ZigbeeDevice: 00000000-1111-2222-3333-000000000001 [0x0001]> received command: {"component":"main","args":{"sensitivity":"Low"},"command":"setSensitivityAdjustment","capability":"stse.sensitivityAdjustment","positional_args":["Low"]}
TRACE || Found CapabilityCommandDispatcher handler in zigbee-smoke-detector -> Aqara Gas Detector Handler
INFO  || <ZigbeeDevice: 00000000-1111-2222-3333-000000000001 [0x0001]> sending Zigbee message: < ZigbeeMessageTx || Uint16: 0x0000, < AddressHeader || src_addr: 0x0000, src_endpoint: 0x01, dest_addr: 0x0001, dest_endpoint: 0x01, profile: 0x0104, cluster: 0xFCC0 >, < ZCLMessageBody || < ZCLHeader || frame_ctrl: 0x04, mfg_code: 0x115F, seqno: 0x00, ZCLCommandId: 0x02 >, < WriteAttribute || < AttributeRecord || attr_id: 0x010C, DataType: Uint8, data: 0x01 > > > >
Failed with message:
Zigbee message channel send was given unexpected message:
ZigbeeMessageTx:
        Uint16: 0x0000
        AddressHeader:
            src_addr: 0x0000
            src_endpoint: 0x01
            dest_addr: 0x0001
            dest_endpoint: 0x01
            profile: 0x0104
            cluster: 0xFCC0
        ZCLMessageBody:
            ZCLHeader:
                frame_ctrl: 0x04
                mfg_code: 0x115F
                seqno: 0x00
                ZCLCommandId: 0x02
            WriteAttribute:
                AttributeRecord:
                    attr_id: 0x010C
                    DataType: Uint8
                    data: 0x01
FAILED
`

Reply:
You need to queue a report of the `PRIVATE_SENSITIVITY_ADJUSTMENT_ATTRIBUTE_ID` so that an event will be generated and the value in the device cache will be correct.

Reply:
@Hongming6 

test.register_coroutine_test(
  "Capability on command should be handled : setSensitivityAdjustment Low",
  function()
    local attr_report_data = {
      { PRIVATE_SENSITIVITY_ADJUSTMENT_ATTRIBUTE_ID, data_types.Uint8.ID, 0x01 }
    }
    test.socket.capability:__queue_receive({ mock_device.id,
      { capability = sensitivityAdjustmentId, component = "main", command = "setSensitivityAdjustment", args = {"Low"}}} )
    test.socket.zigbee:__expect_send({ mock_device.id,
      cluster_base.write_manufacturer_specific_attribute(mock_device, PRIVATE_CLUSTER_ID,
      PRIVATE_SENSITIVITY_ADJUSTMENT_ATTRIBUTE_ID, MFG_CODE, data_types.Uint8, 0x01) })
    test.wait_for_events()

    test.socket.zigbee:__queue_receive({
      mock_device.id,
      zigbee_test_utils.build_attribute_report(mock_device, PRIVATE_CLUSTER_ID, attr_report_data, MFG_CODE)})
    test.socket.capability:__expect_send(mock_device:generate_test_message("main",
      sensitivityAdjustment.sensitivityAdjustment.Low()))
    test.wait_for_events()

    test.socket.capability:__queue_receive({ mock_device.id,
      { capability = sensitivityAdjustmentId, component = "main", command = "setSensitivityAdjustment", args = {"Low"}}} )
    test.socket.capability:__expect_send(mock_device:generate_test_message("main",
      sensitivityAdjustment.sensitivityAdjustment.Low()))
  end
)

BTW.
Please pay attention to the code format. (incorrect indent, extra blank lines, united code style)
Also rebase your branch on time.

Reply:
Thank you, it has been modified